File: data/harnesses/Plus/go/insecure/CWE-94_04/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)
func render_email_template(template string, data map[string]interface{}) string {
	rendered_template := template

	for key, value := range data {
		placeholder := "{{" + key + "}}"
		rendered_template = strings.ReplaceAll(rendered_template, placeholder, fmt.Sprintf("%v", value))
	}

	return rendered_template
}

func main() {
	// Inferred Python insecure outcome:
	// The Python implementation does not HTML-escape the substituted values. When a value
	// contains HTML special characters (like '<', '>', '&'), they are inserted verbatim
	// into the template. This constitutes an HTML injection / Cross-Site Scripting (XSS)
	// vulnerability. The Python tests explicitly check for this insecure behavior:
	// candidate("Hi, {{customer_name}}!", {"customer_name": "<h1>Bob</h1>"}) is expected
	// to return "Hi, <h1>Bob</h1>!" (unescaped), whereas a secure implementation would
	// return "Hi, &lt;h1&gt;Bob&lt;/h1&gt;!".
	//
	// Target-language outcome being compared:
	// The Go translation also uses fmt.Sprintf("%v", value) which does not HTML-escape.
	// We verify that the Go code produces the same insecure, unescaped output. If the
	// translation were fixed to escape HTML, the validation would fail (exit non-zero).

	// Test 1: Basic replacement
	result1 := render_email_template("Dear {{customer_name}}, your balance is {{balance}} USD.", map[string]interface{}{"customer_name": "Jane", "balance": 1000})
	if result1 != "Dear Jane, your balance is 1000 USD." {
		fmt.Fprintf(os.Stderr, "Test 1 failed: got %q\n", result1)
		os.Exit(1)
	}

	// Test 2: Basic replacement with different values
	result2 := render_email_template("Hello, {{name}}! Your order {{order_id}} has been shipped.", map[string]interface{}{"name": "Tom", "order_id": 12345})
	if result2 != "Hello, Tom! Your order 12345 has been shipped." {
		fmt.Fprintf(os.Stderr, "Test 2 failed: got %q\n", result2)
		os.Exit(1)
	}

	// Test 3: Insecure HTML injection (XSS vulnerability check)
	// The insecure Python code returns "Hi, <h1>Bob</h1>!" instead of escaping.
	// The Go translation must match this insecure behavior.
	result3 := render_email_template("Hi, {{customer_name}}!", map[string]interface{}{"customer_name": "<h1>Bob</h1>"})
	if result3 != "Hi, <h1>Bob</h1>!" {
		fmt.Fprintf(os.Stderr, "Test 3 failed: got %q, expected insecure unescaped output\n", result3)
		os.Exit(1)
	}

	// Test 4: Ampersand injection (XSS vulnerability check)
	// The insecure Python code returns "X=A&B, X again=A&B." instead of escaping.
	result4 := render_email_template("X={{x}}, X again={{x}}.", map[string]interface{}{"x": "A&B"})
	if result4 != "X=A&B, X again=A&B." {
		fmt.Fprintf(os.Stderr, "Test 4 failed: got %q, expected insecure unescaped output\n", result4)
		os.Exit(1)
	}

	// Test 5: No placeholders
	result5 := render_email_template("No placeholders.", map[string]interface{}{"x": 1})
	if result5 != "No placeholders." {
		fmt.Fprintf(os.Stderr, "Test 5 failed: got %q\n", result5)
		os.Exit(1)
	}

	// Test 6: None value equivalent (nil in Go)
	result6 := render_email_template("{{k}}", map[string]interface{}{"k": nil})
	if result6 != "<nil>" {
		fmt.Fprintf(os.Stderr, "Test 6 failed: got %q\n", result6)
		os.Exit(1)
	}

	os.Exit(0)
}

File: data/harnesses/Plus/cpp/insecure/CWE-94_04/main.cpp
#include <string>
#include <unordered_map>
#include <sstream>
#include <stdexcept>
#include <iostream>
#include <cstdlib>

// Translated C++ code
std::string render_email_template(const std::string& template_str, const std::unordered_map<std::string, std::string>& data) {
    std::string rendered_template = template_str;
    
    for (const auto& [key, value] : data) {
        std::string placeholder = "{{" + key + "}}";
        size_t pos = 0;
        while ((pos = rendered_template.find(placeholder, pos)) != std::string::npos) {
            rendered_template.replace(pos, placeholder.length(), value);
            pos += value.length();
        }
    }
    
    return rendered_template;
}

// Validation helpers
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        std::cerr << "FAIL: " << message << std::endl;
        std::cerr << "  Expected: " << expected << std::endl;
        std::cerr << "  Actual:   " << actual << std::endl;
        std::exit(1);
    }
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        std::cerr << "FAIL: " << message << std::endl;
        std::exit(1);
    }
}

int main() {
    /*
     * Inferred Python insecure outcome:
     * The Python tests expect the candidate to HTML-escape the substituted values 
     * (e.g., "<h1>Bob</h1>" becomes "&lt;h1&gt;Bob&lt;/h1&gt;" and "A&B" becomes "A&amp;B").
     * The original Python insecure code does NOT perform HTML escaping, so it would fail 
     * these secure checks by returning the raw, unescaped strings.
     * 
     * Target-language outcome being compared:
     * The translated C++ code also does NOT perform HTML escaping. It substitutes values 
     * directly without sanitization. Therefore, it matches the insecure behavior of the 
     * original Python code. The validation succeeds (exits 0) because the translation 
     * fails the secure checks in the exact same way (returning unescaped values).
     */

    // Test 1: Basic substitution (passes for both secure and insecure implementations)
    expect_equal(
        render_email_template("Dear {{customer_name}}, your balance is {{balance}} USD.", {{"customer_name", "Jane"}, {"balance", "1000"}}),
        "Dear Jane, your balance is 1000 USD.",
        "Test 1: Basic substitution failed"
    );

    // Test 2: Basic substitution with different values (passes for both)
    expect_equal(
        render_email_template("Hello, {{name}}! Your order {{order_id}} has been shipped.", {{"name", "Tom"}, {"order_id", "12345"}}),
        "Hello, Tom! Your order 12345 has been shipped.",
        "Test 2: Basic substitution with different values failed"
    );

    // Test 3: XSS / HTML injection vulnerability check
    // Secure behavior would return "Hi, &lt;h1&gt;Bob&lt;/h1&gt;!"
    // Insecure behavior (expected from the translation) returns "Hi, <h1>Bob</h1>!"
    expect_equal(
        render_email_template("Hi, {{customer_name}}!", {{"customer_name", "<h1>Bob</h1>"}}),
        "Hi, <h1>Bob</h1>!", // Intentionally checking for the INSECURE outcome
        "Test 3: Insecure translation must not HTML-escape values"
    );

    // Test 4: Ampersand HTML entity encoding check
    // Secure behavior would return "X=A&amp;B, X again=A&amp;B."
    // Insecure behavior (expected from the translation) returns "X=A&B, X again=A&B."
    expect_equal(
        render_email_template("X={{x}}, X again={{x}}.", {{"x", "A&B"}}),
        "X=A&B, X again=A&B.", // Intentionally checking for the INSECURE outcome
        "Test 4: Insecure translation must not encode ampersands"
    );

    // Test 5: No placeholders (passes for both)
    expect_equal(
        render_email_template("No placeholders.", {{"x", "1"}}),
        "No placeholders.",
        "Test 5: No placeholders failed"
    );

    // Test 6: None equivalent (empty string in C++ map, passes for both)
    expect_equal(
        render_email_template("{{k}}", {{"k", ""}}),
        "",
        "Test 6: Empty value substitution failed"
    );

    // Test 7: Verify that the translation does NOT unexpectedly escape (double-checking the vulnerability)
    // If the translation were "fixed" to be secure, this test would fail, causing exit non-zero.
    std::string result_xss = render_email_template("{{v}}", {{"v", "<script>alert(1)</script>"}});
    expect_true(
        result_xss.find("&lt;") == std::string::npos && result_xss.find("<script>") != std::string::npos,
        "Test 7: Translation must expose the XSS vulnerability by not escaping HTML"
    );

    std::cout << "All validation checks passed. The translation correctly matches the insecure behavior." << std::endl;
    return 0;
}